src/silo/storage/column/insertion_index.cpp

```diff
@@ -0,0 +1,235 @@
+   const auto& insertion = position_and_insertion[1];
+   return std::make_pair(position, insertion);
+}
+
+struct ThreeMerHash {
+   size_t operator()(const InsertionIndex::three_mer_t& three_mer) const {
+      size_t seed = 0;
+      for (const auto one_mer : three_mer) {
+         boost::hash_combine(seed, std::hash<NUCLEOTIDE_SYMBOL>{}(one_mer));
+      }
+      return seed;
+   }
+};
+
+std::vector<InsertionIndex::three_mer_t> extractThreeMers(const std::string& search_pattern) {
+   std::unordered_set<InsertionIndex::three_mer_t, ThreeMerHash> result;
+   for (const auto& continuous_string : splitBy(search_pattern, REGEX_ANY)) {
+      auto continuous_symbols = stringToNucleotideSymbolVector(continuous_string);
+      if (continuous_symbols == std::nullopt) {
+         throw std::runtime_error("Wrong symbol in pattern: " + continuous_string);
+      }
+      for (size_t i = 0; (i + 2) < continuous_string.size(); i += 3) {
+         InsertionIndex::three_mer_t const three_mer{
+            continuous_symbols->at(i),
+            continuous_symbols->at(i + 1),
+            continuous_symbols->at(i + 2)};
+         result.insert(three_mer);
+      }
+   }
+   return {result.begin(), result.end()};
+}
+
+}  // namespace
+
+void InsertionIndex::InsertionPosition::buildThreeMerIndex() {
+   using bitset_one_mers_t = NucleotideSymbolMap<bool>;
+   using bitset_two_mers_t = NucleotideSymbolMap<bitset_one_mers_t>;
+   using bitset_three_mers_t = NucleotideSymbolMap<bitset_two_mers_t>;
+
+   for (size_t insertion_id = 0; insertion_id < insertions.size(); ++insertion_id) {
+      const auto& insertion = insertions[insertion_id];
+      const auto& insertion_value = insertion.value;
+
+      if (insertion_value.size() < 3) {
+         continue;
+      }
+
+      const auto opt_nuc_symbol_ids = stringToNucleotideSymbolVector(insertion_value);
+      if (opt_nuc_symbol_ids == std::nullopt) {
+         throw silo::PreprocessingException("Illegal character in insertion: " + insertion_value);
+      }
+      const auto& nuc_symbol_ids = *opt_nuc_symbol_ids;
+
+      bitset_three_mers_t unique_three_mers{};
+      for (size_t i = 0; i < (nuc_symbol_ids.size() - 2); ++i) {
+         unique_three_mers[nuc_symbol_ids[i]][nuc_symbol_ids[i + 1]][nuc_symbol_ids[i + 2]] = true;
+      }
+
+      for (const NUCLEOTIDE_SYMBOL symbol1 : NUC_SYMBOLS) {
+         for (const NUCLEOTIDE_SYMBOL symbol2 : NUC_SYMBOLS) {
+            for (const NUCLEOTIDE_SYMBOL symbol3 : NUC_SYMBOLS) {
+               if (unique_three_mers[symbol1][symbol2][symbol3]) {
+                  three_mer_index[symbol1][symbol2][symbol3].push_back(insertion_id);
+               }
+            }
+         }
+      }
+   }
+}
+
+std::unique_ptr<InsertionIndex::sequence_ids_t> InsertionIndex::InsertionPosition::
+   searchWithThreeMerIndex(
+      const std::vector<three_mer_t>& search_three_mers,
+      const std::regex& search_pattern
+   ) const {
+   assert(!search_three_mers.empty());
```

Comment:
Please don't use `assert`. We've had those already. This aborts the program if it fails. SILO is a continuously running program (currently a REST API), you don't want to shut it down. You want to throw exceptions that you can handle, so that you can send subsequent requests.

Reply:
If compiled in release mode asserts are actually no-ops (i.e. function purely as sanity checks in debugging)

Reply:
I left the assert for now because it should not affect a running instance compiled in release mode. But if you prefer to not have `assert`s In the codebase, then I will remove it.

Reply:
I don't really see the point of having it. You could as well throw an exception.

The check is done on calling side, too. So we could as well omit it either here or on the calling side, whichever makes more sense.

Reply:
Alright. I removed it 👍